### PR TITLE
DINOv2 Losses

### PR DESCRIPTION
**Pull request by yutong-xiang-97**
## What has changed and why?

We carry the `DINOLoss` and `iBOTPatchLoss` from original DINOv2. The only two differences are that:

- Import xFormers' cross entropy only if `XFORMERS_ENABLED` is True
- Use `dist.is_initialized()` to control the `all_reduce` operation of `B` in distributed setting in the `iBOTPatchLoss`'s `sinkhorn_knopp_teacher` (line 173-174).

We don't need the `KoleoLoss` cuz it is identical of the one in LightlySSL, and the latter handles the situation when the cos sim is -1 correctly.

Tests cover the correctness of:
- two centering methods
- center momentum
- loss values

## How has it been tested?

Running unit tests.

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_methods/dinov2/dinov2_loss.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        # WARNING:
+        #   as self.center is a float32, everything gets casted to float32 afterwards
+        #
+        # teacher_patch_tokens = teacher_patch_tokens.float()
+        # return F.softmax((teacher_patch_tokens.sub_(self.center.to(teacher_patch_tokens.dtype))).mul_(1 / teacher_temp), dim=-1)
+
+        return F.softmax((teacher_patch_tokens - self.center) / teacher_temp, dim=-1)
+
+        # this is experimental, keep everything in float16 and let's see what happens:
+        # return F.softmax((teacher_patch_tokens.sub_(self.center)) / teacher_temp, dim=-1)
```
Comment (JonasWurst) [inline src/lightly_train/_methods/dinov2/dinov2_loss.py]: how are we going to handle this? --> follow up issue?

File: src/lightly_train/_methods/dinov2/dinov2_loss.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+            _t = self.async_batch_center / (self.len_teacher_output * world_size)
+
+            self.center = self.center * self.center_momentum + _t * (
+                1 - self.center_momentum
+            )
+
+            self.updated = True
+
+
+class iBOTPatchLoss(nn.Module):
```
Comment (JonasWurst) [inline src/lightly_train/_methods/dinov2/dinov2_loss.py]:
```suggestion
class IBOTPatchLoss(nn.Module):
```

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+# All rights reserved.
+#
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+#
+import pytest
+import torch
+import torch.distributed as dist
+
+from lightly_train._methods.dinov2.dinov2_loss import DINOLoss, iBOTPatchLoss
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]:
```suggestion
from lightly_train._methods.dinov2.dinov2_loss import DINOLoss, IBOTPatchLoss
```

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+@pytest.mark.usefixtures("no_dist")
+class TestDINOLoss:
+    def test_softmax_center_teacher(
+        self,
+        batch_size=4,
+        out_dim=2,
+        teacher_temp=0.04,
+        student_temp=0.1,
+        center_momentum=0.9,
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]: we are not varying nor reusing the function --> can be set inside

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        assert torch.allclose(sums, torch.ones(batch_size))
+
+    def test_sinkhorn_knopp_teacher(
+        self,
+        batch_size=4,
+        out_dim=2,
+        teacher_temp=0.04,
+        student_temp=0.1,
+        center_momentum=0.9,
+        n_iterations=4,
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]: similar here

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        row_sums = Q.sum(dim=1)
+        assert torch.allclose(row_sums, torch.ones(batch_size))
+
+    def test_update_center_momentum(
+        self,
+        batch_size=4,
+        out_dim=2,
+        student_temp=0.1,
+        center_momentum=0.9,
+        mean=2,
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]: similar

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+        expected_center = mean * (1 - center_momentum) * torch.ones(out_dim)
+        assert torch.allclose(dino_loss.center, expected_center)
+
+    def test_forward(
+        self,
+        out_dim=2,
+        teacher_temp=0.04,
+        student_temp=0.1,
+        center_momentum=0.9,
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]: .

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+        patch_out_dim=2,
+        teacher_temp=0.04,
+        student_temp=0.1,
+        center_momentum=0.9,
+    ) -> None:
+        """Test that the softmax_center_teacher method returns a tensor
+        with the same shape as the input tensor and that each row sums to 1.
+        """
+
+        ibot_loss = iBOTPatchLoss(
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]:
```suggestion
        ibot_loss = IBOTPatchLoss(
```

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+@pytest.mark.usefixtures("no_dist")
+class TestIBotPatchLoss:
+    def test_softmax_center_teacher(
+        self,
+        batch_size=4,
+        patch_out_dim=2,
+        teacher_temp=0.04,
+        student_temp=0.1,
+        center_momentum=0.9,
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]: .

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        n_iterations=4,
+    ) -> None:
+        """Test that the sinkhorn_knopp_teacher method returns a tensor
+        with the same shape as the input tensor and that each row sums to 1.
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py:130]: column also checked

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        self,
+        batch_size=4,
+        out_dim=2,
+        teacher_temp=0.04,
+        student_temp=0.1,
+        center_momentum=0.9,
+        n_iterations=4,
+    ) -> None:
+        """Test that the sinkhorn_knopp_teacher method returns a tensor
+        with the same shape as the input tensor and that each row sums to 1.
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py:44]: column also checked

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+        assert torch.allclose(sums, torch.ones(batch_size))
+
+    def test_sinkhorn_knopp_teacher(
+        self,
+        batch_size=4,
+        patch_out_dim=2,
+        teacher_temp=0.04,
+        student_temp=0.1,
+        center_momentum=0.9,
+        n_iterations=4,
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]: .

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        row_sums = Q.sum(dim=1)
+        assert torch.allclose(row_sums, torch.ones(batch_size))
+
+    def test_update_center_momentum(
+        self,
+        batch_size=4,
+        patch_out_dim=2,
+        student_temp=0.1,
+        center_momentum=0.9,
+        mean=2,
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]: .

File: tests/_methods/dinov2/test_dinov2_loss.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+        expected_center = mean * (1 - center_momentum) * torch.ones(patch_out_dim)
+        assert torch.allclose(ibot_loss.center, expected_center)
+
+    def test_forward_masked(
+        self,
+        out_dim=2,
+        teacher_temp=0.1,
+        student_temp=0.2,
+        center_momentum=0.9,
```
Comment (JonasWurst) [inline tests/_methods/dinov2/test_dinov2_loss.py]: .

File: pyproject.toml [hunk 15]
```diff
@@ -121,7 +121,9 @@ ignore_missing_imports = true
  
 [[tool.mypy.overrides]]
 module = [
-    "lightly_train._modules.teachers.dinov2.*"
+    "lightly_train._modules.teachers.dinov2.*",
+    "lightly_train._methods.dinov2.dinov2_loss.*",
+    "tests._methods.dinov2.test_dinov2_loss.*"
```
Comment (JonasWurst) [inline pyproject.toml]: why is the test excluded?

**Review by JonasWurst — COMMENTED** [review]
Small issues only.

In general:
- add typing
- cleanup dead code comments
- create follow up issues for problematic section

File: tests/_methods/dinov2/test_dinov2_loss.py (hunk 11 above)
Comment (yutong-xiang-97) [reply tests/_methods/dinov2/test_dinov2_loss.py:44]: Took a closer look and the statement for columns is not correct. Removed.

File: tests/_methods/dinov2/test_dinov2_loss.py (hunk 10 above)
Comment (yutong-xiang-97) [reply tests/_methods/dinov2/test_dinov2_loss.py:130]: Took a closer look and the statement for columns is not correct. Removed.

File: src/lightly_train/_methods/dinov2/dinov2_loss.py (hunk 1 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_methods/dinov2/dinov2_loss.py]: This is directly carried from the original dino loss and it never got fixed. So I presume it's not a big deal and doesn't affect the final result significantly. My opinion is to leave it as is.

File: pyproject.toml (hunk 15 above)
Comment (yutong-xiang-97) [reply pyproject.toml]: It was not typed. Now it is fixed.

File: tests/_methods/dinov2/test_dinov2_loss.py (hunk 11 above)
Comment (JonasWurst) [reply tests/_methods/dinov2/test_dinov2_loss.py:44]: what exactly was wrong?

File: src/lightly_train/_methods/dinov2/dinov2_loss.py (hunk 1 above)
Comment (JonasWurst) [reply src/lightly_train/_methods/dinov2/dinov2_loss.py]: yes, that is clear. I'm thinking about possible implications if a certain precision is requested.

two options:
- If you are sure that there are no implications, we can just ignore all the comments, remove them and resolve.
- If there are possible implications or you are not sure about: lets create an issue to clarify this, remove the comments and resolve

File: src/lightly_train/_methods/dinov2/dinov2_loss.py [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+
+    def forward(
+        self,
+        student_output_list: List[Tensor],
+        teacher_out_softmaxed_centered_list: List[Tensor],
+    ) -> Tensor:
+        """
+        Cross-entropy between softmax outputs of the teacher and student networks.
+        """
+        # TODO: Use cross_entropy_distribution here
```
Comment (JonasWurst) [inline src/lightly_train/_methods/dinov2/dinov2_loss.py]:
```suggestion
```

File: src/lightly_train/_methods/dinov2/dinov2_loss.py [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+        super().__init__()
+        self.student_temp = student_temp
+        self.center_momentum = center_momentum
+        self.register_buffer("center", torch.zeros(1, 1, patch_out_dim))
+        # Type hint for self.center
+        self.center: torch.Tensor
+        self.updated = True
+        self.reduce_handle = None
+        # self.len_teacher_patch_tokens = None
+        # self.async_batch_center = None
```
Comment (JonasWurst) [inline src/lightly_train/_methods/dinov2/dinov2_loss.py]:
```suggestion
```

File: src/lightly_train/_methods/dinov2/dinov2_loss.py [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+    @torch.no_grad()
+    def sinkhorn_knopp_teacher(
+        self,
+        teacher_output: Tensor,
+        teacher_temp: float,
+        n_masked_patches_tensor: Tensor,
+        n_iterations: int = 3,
+    ) -> Tensor:
+        teacher_output = teacher_output.float()
+        # world_size = dist.get_world_size() if dist.is_initialized() else 1
```
Comment (JonasWurst) [inline src/lightly_train/_methods/dinov2/dinov2_loss.py]:
```suggestion
```

File: src/lightly_train/_methods/dinov2/dinov2_loss.py [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+        Q = torch.exp(
+            teacher_output / teacher_temp
+        ).t()  # Q is K-by-B for consistency with notations from our paper
+        # B = Q.shape[1] * world_size # number of samples to assign
```
Comment (JonasWurst) [inline src/lightly_train/_methods/dinov2/dinov2_loss.py]:
```suggestion
```

File: src/lightly_train/_methods/dinov2/dinov2_loss.py [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+        self,
+        student_patch_tokens_masked: Tensor,
+        teacher_patch_tokens_masked: Tensor,
+        student_masks_flat: Tensor,
+        n_masked_patches: int | None = None,
+        masks_weight: Tensor | None = None,
+    ) -> Tensor:
+        t = teacher_patch_tokens_masked
+        s = student_patch_tokens_masked
+        # loss = torch.sum(t * F.log_softmax(s / self.student_temp, dim=-1), dim=-1)
```
Comment (JonasWurst) [inline src/lightly_train/_methods/dinov2/dinov2_loss.py]:
```suggestion
```

**Review by JonasWurst — CHANGES_REQUESTED** [review]
Let's remove the dead code pieces

File: tests/_methods/dinov2/test_dinov2_loss.py (hunk 11 above)
Comment (yutong-xiang-97) [reply tests/_methods/dinov2/test_dinov2_loss.py:44]: There wasn't be any constrains on the sum of the columns. Only the rows should be summed to 1.

File: src/lightly_train/_methods/dinov2/dinov2_loss.py (hunk 1 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_methods/dinov2/dinov2_loss.py]: I don't think there would be an issue with the current solution. Hardly ever will people train with float64 so the tensors can only get upcasted. I removed the dead code.

File: tests/_methods/dinov2/test_dinov2_loss.py (hunk 11 above)
Comment (JonasWurst) [reply tests/_methods/dinov2/test_dinov2_loss.py:44]: Yes, but In my memory, sinkhorn knopp will lead to exactly that value. So the test was actually fine imo
